src/backend/distributed/commands/role.c

```diff
@@ -0,0 +1,221 @@
+#include "postgres.h"
+
+#include "access/heapam.h"
+#include "access/htup_details.h"
+#if PG_VERSION_NUM >= 120000
+#include "access/table.h"
+#endif
+#include "catalog/pg_authid.h"
+#include "distributed/commands/utility_hook.h"
+#include "distributed/deparser.h"
+#include "distributed/master_protocol.h"
+#include "distributed/worker_transaction.h"
+#include "nodes/parsenodes.h"
+#include "utils/builtins.h"
+#include "utils/rel.h"
+#include "utils/syscache.h"
+#include "nodes/makefuncs.h"
+#include "catalog/catalog.h"
+
+List * ProcessAlterRoleStmt(AlterRoleStmt *stmt, const char *queryString);
+List * CreateAlterRoleQueriesForCopyingAllUsers(void);
+void PutEncryptedPasswordInQueryTree(DefElem *option, const char *rolename);
+const char * prepareAlterRoleQueryWithUDF(AlterRoleStmt *stmt);
+
+
+bool EnableAlterRolePropagation = true;
+
+/*
+ * ProcessAlterRoleStmt actually creates the plan we need to execute for alter
+ * role statement.
+ */
+List *
+ProcessAlterRoleStmt(AlterRoleStmt *stmt, const char *queryString)
+{
+	ListCell *optionCell = NULL;
+	List *commands = NIL;
+
+	if (!IsCoordinator())
+	{
+		return NIL;
+	}
+
+	foreach(optionCell, stmt->options)
+	{
+		DefElem *option = (DefElem *) lfirst(optionCell);
+
+		if (EnableAlterRolePropagation && strcasecmp(option->defname, "password") == 0)
+		{
+			PutEncryptedPasswordInQueryTree(option, stmt->role->rolename);
+			break;
+		}
+	}
+
+	commands = list_make1((void *) prepareAlterRoleQueryWithUDF(stmt));
+
+	if (list_length(commands) <= 0)
+	{
+		return NIL;
+	}
+	return NodeDDLTaskList(ALL_WORKERS, commands);
+}
+
+
+const char *
+prepareAlterRoleQueryWithUDF(AlterRoleStmt *stmt)
+{
+	StringInfoData alterRoleQueryBuffer = { 0 };
+
+	initStringInfo(&alterRoleQueryBuffer);
+	appendStringInfo(&alterRoleQueryBuffer,
+					 "SELECT run_utility_query_if_role_exists(%s, %s)",
+					 quote_literal_cstr(stmt->role->rolename),
+					 quote_literal_cstr(DeparseTreeNode((Node *) stmt)));
+
+	return pstrdup(alterRoleQueryBuffer.data);
+}
+
+
+void
+PutEncryptedPasswordInQueryTree(DefElem *option, const char *rolename)
+{
+	Relation pgAuthId = NULL;
+	HeapTuple tuple = NULL;
+	TupleDesc pgAuthIdDescription = NULL;
+	bool isNull = true;
+	Datum passwordDatum;
+
+	pgAuthId = heap_open(AuthIdRelationId, AccessShareLock);
+	tuple = SearchSysCache1(AUTHNAME, CStringGetDatum(rolename));
+
+	pgAuthIdDescription = RelationGetDescr(pgAuthId);
+	passwordDatum = heap_getattr(tuple, Anum_pg_authid_rolpassword,
+								 pgAuthIdDescription, &isNull);
+
+	heap_close(pgAuthId, AccessShareLock);
+	ReleaseSysCache(tuple);
+
+	if (isNull)
+	{
+		option->arg = NULL;
+	}
+	else
+	{
+		((Value *) (option->arg))->val.str = TextDatumGetCString(passwordDatum);
+	}
+}
+
+
+static const char *
+CreateAlterRoleQueryForCopyingUser(HeapTuple tuple)
```

Comment:
-`GenerateAlterRoleIfExistsCommand()`
-  Pass description, do not re-open heap
- Missing comment